### ISSUE TITLE
[Epic] Configurable Avro sink schemas

### ISSUE THREAD

**Issue by umanwizard** [MaterializeInc/materialize#17209]
## Product outcomes

Users can customize the Avro schema that is generated for a Kafka Avro sink. The customizations are limited to the common requests we've gotten to date.

## Definite work items

* [x] https://github.com/MaterializeInc/materialize/issues/21557
* [x] https://github.com/MaterializeInc/materialize/issues/21559
* [x] https://github.com/MaterializeInc/materialize/issues/21829
* [x] https://github.com/MaterializeInc/materialize/issues/21722
* [x] https://github.com/MaterializeInc/materialize/issues/21661

## Possible work items

* [ ] https://github.com/MaterializeInc/materialize/issues/17668

## Descoped work items 

* [ ] https://github.com/MaterializeInc/materialize/issues/21558
* [ ] https://github.com/MaterializeInc/materialize/issues/21662

## Action log

* *22 September 2023.* Add #17668 as a possible work item, as it may be required to properly report errors when forcing non-nullability (#21722).
* *20 September 2023.* Split out https://github.com/MaterializeInc/materialize/issues/21829 from https://github.com/MaterializeInc/materialize/issues/21662 and descoped the latter.
* *18 September 2023.* Added #21722 as a work item.
* *20 October 2023.* Added https://github.com/MaterializeInc/materialize/issues/21661 to epic

----

*Original issue for posterity.*

Currently, Materialize itself generates the schema for an Avro sink, and does not allow any user customization. We've received a few requests along these lines, for example

(1) the ability to add `doc` parameters to various fields, which are interpreted by downstream systems,
(2) the ability to control which of the various Avro timestamp logical types is emitted

And probably more.

There is some design space on how this could work. (How does the user communicate their choices to us? How much freedom do we allow?)

**Comment by sjwiesman** [comment]
Since this is titled musings, I'm going to do just that. 

It seems like for both sources and sinks, whether or not materialize (de/en)codes a type using its logical type should be a format parameter. In theory, I'd love logical types enabled to be the default, but I wonder if we can do that in a backward-compatible way. I'll probably say we can't, and then Nikhil will show up and say we can 🤷 

My gut reaction on doc parameters is that they should be tied to comments somehow. What would be cool is if we parsed out actual sql comments. 

```sql
CREATE MATERIALIZED VIEW to_be_sinked AS 
SELECT
--- doc: I'm an id!
id,
--- doc: I'm a poorly named count
count(*) AS c
FROM foo
```

The other, probably more sensible way, would be to also have it as a parameter on the type in your sink object. 

```sql
CREATE SOURCE avro_source
  FROM KAFKA CONNECTION kafka_connection (TOPIC 'test_topic')
  FORMAT AVRO [something about doc comments here] USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_connection
  WITH (SIZE = '3xsmall');
```

**Comment by benesch** [comment]
> I'll probably say we can't, and then Nikhil will show up and say we can

I also agree that adding a logical type to the schema is not backwards compatible! Will need some sort of `FORMAT AVRO (GOOD DECIMAL)` option. (Better name pending.)

Re doc parameters: PostgreSQL has the notion of comments already! See this [SO answer](https://stackoverflow.com/a/32071293). You can add a comment to a column, even.

I think the path forward here is pretty clear. Add support for `COMMENT` to Materialize. In the same PR, teach the Avro schema generator to automatically add `doc` fields when generating an Avro schema for any columns with comments. This will be backwards compatible, because it will only change the behavior when creating sinks on new views that have comments.

Actually, one thing that's maybe a little weird: comments can be changed without dropping/recreating the underlying view. That means the schema that we generate will change after restart to reflect the new comments. Maybe that's ok?

**Comment by sjwiesman** [comment]
I’d say modifying a schemas doc is both backwards and forwards compatible

**Comment by umanwizard** [comment]
IMO any change for which schema resolution is still possible according to the Avro schema resolution rules should be considered compatible. This includes modifying schema doc fields.

**Comment by moulimukherjee** [comment]
Should https://github.com/MaterializeInc/materialize/issues/21661 be added here as well?

**Comment by benesch** [comment]
> Should #21661 [Make it easier to see the Avro schema that would be created for an Avro sink](https://github.com/MaterializeInc/materialize/issues/21661) be added here as well?

Discussed on Slack. Plan is to keep this epic focused on the set of things strictly required to unblock the customer's use case, but we've linked #21661 into https://github.com/MaterializeInc/product/issues/225 as a future nice to have.
